src/bun.js/bindings/BunProcess.cpp

```diff
@@ -655,121 +734,173 @@ static void onDidChangeListeners(EventEmitter& eventEmitter, const Identifier& e
     }
 
     if (auto signalNumber = signalNameToNumberMap->get(eventName.string())) {
+#if !OS(WINDOWS)
         if (signalNumber != SIGKILL && signalNumber != SIGSTOP) {
+#else
+        if (signalNumber != SIGKILL) {
+#endif
             uint32_t contextId = eventEmitter.scriptExecutionContext()->identifier();
             Locker lock { signalToContextIdsMapLock };
 
             if (isAdded) {
                 if (!signalToContextIdsMap->contains(signalNumber)) {
-                    HashSet<uint32_t> contextIds;
-                    contextIds.add(contextId);
-                    signalToContextIdsMap->set(signalNumber, contextIds);
-
-                    lock.unlockEarly();
-
+                    SignalHandleValue signal_handle = {
+#if OS(WINDOWS)
+                        .handle = nullptr,
+#endif
+                        .contextIds = {},
+                    };
+                    signal_handle.contextIds.add(contextId);
+#if !OS(WINDOWS)
                     struct sigaction action;
                     memset(&action, 0, sizeof(struct sigaction));
 
                     // Set the handler in the action struct
-                    action.sa_handler = [](int signalNumber) {
-                        if (UNLIKELY(signalNumberToNameMap->find(signalNumber) == signalNumberToNameMap->end()))
-                            return;
-
-                        Locker lock { signalToContextIdsMapLock };
-                        if (UNLIKELY(signalToContextIdsMap->find(signalNumber) == signalToContextIdsMap->end()))
-                            return;
-                        auto contextIds = signalToContextIdsMap->get(signalNumber);
-
-                        for (int contextId : contextIds) {
-                            auto* context = ScriptExecutionContext::getScriptExecutionContext(contextId);
-                            if (UNLIKELY(!context))
-                                continue;
-
-                            JSGlobalObject* lexicalGlobalObject = context->jsGlobalObject();
-                            Zig::GlobalObject* globalObject = jsCast<Zig::GlobalObject*>(lexicalGlobalObject);
-
-                            Process* process = jsCast<Process*>(globalObject->processObject());
-
-                            context->postCrossThreadTask(*process, &Process::emitSignalEvent, signalNumber);
-                        }
-                    };
+                    action.sa_handler = signalHandler;
 
                     // Clear the sa_mask
                     sigemptyset(&action.sa_mask);
                     sigaddset(&action.sa_mask, signalNumber);
                     action.sa_flags = SA_RESTART;
 
                     sigaction(signalNumber, &action, nullptr);
+#else
+                    // To be more consistant, make sure the loop used is always the main thread's loop
+                    if (eventEmitter.scriptExecutionContext()->isMainThread()) {
+                        signal_handle.handle = Bun__UVSignalHandle__init(
```

Comment:
Can you make this a field on the class which is constructed/destructed, instead of manually initialized and uninitialized? Since we're in C++ we can do that and this code will be duplicated less + safer